packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -447,8 +447,30 @@ const apisRunningById = new Map()
 const apisRunningByTraceId = new Map()
 let waitingForCasacadeToFinish = []
 
-module.exports = async (api, args = {}, { pluginSource, activity } = {}) =>
-  new Promise(resolve => {
+module.exports = (api, args = {}, { pluginSource, activity } = {}) => {
+  const plugins = store.getState().flattenedPlugins
+
+  // Get the list of plugins that implement this API.
+  // Also: Break infinite loops. Sometimes a plugin will implement an API and
+  // call an action which will trigger the same API being called.
+  // `onCreatePage` is the only example right now. In these cases, we should
+  // avoid calling the originating plugin again.
+  let implementingPlugins = plugins.filter(
+    plugin => plugin.nodeAPIs.includes(api) && plugin.name !== pluginSource
+  )
+
+  if (api === `sourceNodes` && args.pluginName) {
+    implementingPlugins = implementingPlugins.filter(
+      plugin => plugin.name === args.pluginName
+    )
+  }
+
+  // If there's no implementing plugins, return early.
+  if ((implementingPlugins.length === 0) & !args.deferNodeMutation) {
```

Comment:
without the `deferNodeMutation` check, the develop server gets stuck on starting webpack 🤷

Reply:
This is confusing and I'd prefer to understand why this cause server to get stuck (and fix that) instead. It does look like we rely on some undefined timing?

Reply:
Hmm, how was it getting stuck? I'm testing it locally right now and I can't reproduce that after commenting out `deferNodeMutation` part of check 🤔

Reply:
ok, I see this is happening in themes e2e tests, so will try that